app/models/concerns/host_params.rb

```diff
@@ -8,59 +8,65 @@ module HostParams
     include ParameterValidators
     attr_reader :cached_host_params
 
-    def params
-      host_params.update(lookup_keys_params)
-    end
+    register_info_provider :host_params_info
+  end
 
-    def clear_host_parameters_cache!
-      @cached_host_params = nil
-    end
+  def host_params_info
+    { 'parameters' => params }
+  end
 
-    def host_inherited_params(include_source = false)
-      hp = {}
-      params = host_inherited_params_objects
-      params.each do |param|
-        source = param.associated_type
-        options = {:value => param.value,
-                   :source => source,
-                   :safe_value => param.safe_value }
-        if source != 'global'
-          options.merge!(:source_name => param.associated_label)
-        end
-        hp.update(Hash[param.name => include_source ? options : param.value])
-      end
-      hp
-    end
+  def params
+    host_params
+  end
 
-    def host_params
-      return cached_host_params unless cached_host_params.blank?
-      hp = host_inherited_params
-      # and now read host parameters, override if required
-      host_parameters.authorized(:view_params).each { |p| hp[p.name] = p.value }
-      @cached_host_params = hp
-    end
+  def clear_host_parameters_cache!
+    @cached_host_params = nil
+  end
 
-    def host_inherited_params_objects
-      params = CommonParameter.all
-      if SETTINGS[:organizations_enabled] && organization
-        params += extract_params_from_object_ancestors(organization)
+  def host_inherited_params(include_source = false)
+    hp = {}
+    params = host_inherited_params_objects
+    params.each do |param|
+      source = param.associated_type
+      options = {:value => param.value,
+                 :source => source,
+                 :safe_value => param.safe_value }
+      if source != 'global'
+        options.merge!(:source_name => param.associated_label)
       end
+      hp.update(Hash[param.name => include_source ? options : param.value])
+    end
+    hp
+  end
 
-      if SETTINGS[:locations_enabled] && location
-        params += extract_params_from_object_ancestors(location)
-      end
+  def host_params
+    return cached_host_params unless cached_host_params.blank?
+    hp = host_inherited_params
+    # and now read host parameters, override if required
+    host_parameters.authorized(:view_params).each { |p| hp[p.name] = p.value }
+    @cached_host_params = hp
```

Comment:
Sorry, I don't follow that logic. How is that different now? In the current case:

1. User gets `@host` in the UI - cache is created if we call host_params 
2. User changes hostgroup -> controller calls `refresh_host` which fetches params from cache - values are wrong.
3. They are *not* wrong because the `@host` object is regenerated.. but we created a nearly useless cache.

In the case I propose:

1. User gets `@host` in the UI - cache is created if we call host_params. Let's save it as `Rails.cache.write("host_params_#{host.fqdn}", host_params)'.
2. User changes hostgroup -> controller calls `refresh_host` which fetches params from cache - values are wrong. At this point (https://github.com/theforeman/foreman/blob/develop/app/controllers/hosts_controller.rb#L733) we should call `@host.clear_host_parameters_cache!` and empty the cache. Only after that and after actually submitting changes to a host we should call that method (I don't know why that's not the case now, probably broken)
3. They are *not* wrong in any case now, no matter any instance they hit, and they're reusable. 

Anyway, I'm realizing it's maybe too much work to do in this PR, give me a nudge at some point and I'll do it separately